python/ray/exceptions.py

```diff
@@ -243,7 +246,27 @@ def __str__(self):
             return ("The actor died because of an error" +
                     " raised in its creation task, " +
                     self.creation_task_error.__str__())
-        return ("The actor died unexpectedly before finishing this task.")
+
+        base_error_msg = (
+            "The actor died unexpectedly before finishing this task")
+        if self.cause:
+            if self.cause.HasField("worker_died_context"):
```

Comment:
I think alternative (probably the cleaner approach) approach is to have a different exception type per each exception (all of them inherited from ActorDiedError).

And then we can write a documentation that explains each of failure.

Reply:
Any thought on this?

Reply:
if we remove enum it should be simpler...

Reply:
Yeah maybe we remove the enum? Seems like a lot of code complexity when we could have instead stored the message as {"error_class": "WorkerDied", "message": the actual message}.

Reply:
It is actually not enum, but oneof. 

I made it have oneof because it will have more metadata per error type (that might differ) in the future. For example, we probably want to include why workers are dead when actor is dead by worker died. 

But alternatively, I can just construct the complete error message from GCS and not include metadata at all. That will certainly simplify the code here. (and in this code, we just print error_message field). Let me try with this approach.

Also, what do you guys think about exception types? Do you prefer to have 1 exception ActorDiedError, or `class ActorOutOfScope(ActorDied)` error?